Add instances, compat

treeowl opened the pull request:
Add instances, compat

* Add `mtl`-style instances. These should be checked for
  reasonableness!

* Finish adding `Eq` and `Ord` instances. Either I missed key
  bits the first time around or they were accidentally removed.

* Add conversions to and from `LogicT` and conversion to arbitrary
  logic monad transformers.

* Require `logict >= 0.7.1.0` for consistent `MonadLogic`
  superclasses.

* Add the beginnings of a test suite. This doesn't yet test
  the `MonadError` instance, and doesn't adequately test the
  `MonadReader` instance.

Starts to address #2
Closes #22
Closes #23

treeowl commented:
The behavior of `local` seems to match that expected in [this resolved `logict` issue](https://github.com/Bodigrim/logict/issues/1), so that's probably fine.

dagit commented:
I added some scaffolding for adding tests last night. Could you turn some of these questions about semantics into tests?

treeowl commented on src/Control/Monad/Logic/Sequence/Internal.hs:
```diff
@@ -305,6 +319,47 @@ observeAll :: Seq a -> [a]
+toMonadLogicT :: (Monad m, Monad (t m), MonadTrans t, Alternative (t m)) => SeqT m a -> t m a
+toMonadLogicT (toView -> m) = lift m >>= \r -> case r of
+  Empty -> empty
+  a :< s -> pure a <|> toMonadLogicT s
```
I'm not a huge fan of the name of this function. Any better ideas? Also, it suggests another function of type

```haskell
otherFunction :: (Monad m, Monad n, Alternative n) => (forall x. m a -> n a) -> SeqT m a -> n a
otherFunction phi (toView -> m) = phi m >>= \r -> case r of
  Empty -> empty
  a :< s -> pure a <|> otherFunction phi s
```
so that
```haskell
toMonadLogicT = otherFunction lift
```

Do we want that too, or is it too specialized?

dagit replied:
I'd probably call the first one `fromSeqT`.  What is the purpose of `forall x`? Did you mean to write `(forall x. m x -> n x)`?

Maybe `otherFunction` could be called `generalize`? I don't have a lot of good ideas on the name for that one. I think `streamly` has something similar but slightly more constrained like that in our types would be like `SeqT m a -> SeqT n a`, but I can't remember what they call it.

Ah found it: https://hackage.haskell.org/package/streamly-0.8.0/docs/Streamly-Internal-Data-Fold.html#v:hoist
```haskell
hoist :: (forall x. m x -> n x) -> Fold m a b -> Fold n a b
```
Your type is more general, but it's a pretty similar concept.

treeowl replied:
Yes, I meant `forall x. m x -> n x`. `hoist` is something we need to consider too, but it's different!

treeowl commented:
@dagit, I don't see any scaffolding, aside from two different test-suite entries in the `.cabal` file. I don't have any experience *using* Hedgehog, though I've made a few contributions to it, and my HSpec is pretty "monkey see, monkey do". If you could maybe add a tiny bit of actual testing code I could surely expand it.

dagit commented:
Yeah, the actual `Test.hs` is pretty empty. I haven't had a chance to populate it yet. Just make sure to steer clear of the do-nothing test suite. That only exists to work around some issues I was having with specifying that I wanted no testing on certain GHC versions (due to the testing libraries not working there).

dagit commented:
This still has some TODOs in the code, so I'm not sure if you're still working on it.

treeowl commented:
I am, but I should have it in better shape by tomorrow or at the latest
Tuesday. Do you have any examples to point to of hedgehog+hspec? The
hedgehog documentation is a mess. Alternatively, I can just use QuickCheck,
which has the advantage of familiarity.

On Sun, Jul 25, 2021, 2:45 PM Jason Dagit ***@***.***> wrote:

> This still has some TODOs in the code, so I'm not sure if you're still
> working on it.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-886242041>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7MC2FAQR57FI4W7AN3TZRLTXANCNFSM5A53S64Q>
> .
>

dagit commented:
Okay. No rush. I just wanted to check before I merged it.

I'm hoping to add some simple test cases soon. Maybe today? Although, I spent a long time on this yesterday and I should probably not work on it all weekend. I mostly picked hedgehog because I think doing generators by type class is a bit of a design flaw. However, quickcheck is welly known and all that and has integration with hspec. So if you prefer it, go ahead. Just know that you might have to fight with the CI a bit as adding dependencies has that effect.

treeowl commented:
I don't particularly prefer it; I just have no idea how to use hedgehog as
yet.

On Sun, Jul 25, 2021, 3:09 PM Jason Dagit ***@***.***> wrote:

> Okay. No rush. I just wanted to check before I merged it.
>
> I'm hoping to add some simple test cases soon. Maybe today? Although, I
> spent a long time on this yesterday and I should probably not work on it
> all weekend. I mostly picked hedgehog because I think doing generators by
> type class is a bit of a design flaw. However, quickcheck is welly known
> and all that and has integration with hspec. So if you prefer it, go ahead.
> Just know that you might have to fight with the CI a bit as adding
> dependencies has that effect.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-886244770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7PQYLF6FR6XT5AFR7DTZROM7ANCNFSM5A53S64Q>
> .
>

treeowl commented:
I'm making progress with Hedgehog. Not so sure it's any easier to understand what's going on with my generators than in QuickCheck, really, but whatever. I think the monad to use for most purposes is

```haskell
newtype TestM a = TestM (Free TestF a)
data TestF = TestF !Int [a]
```

where `Free` is a locally defined free monad. A free monad over a general container type like this tends not to hide any mistakes. It's locally defined because using the one in `free` would force us to write an orphan `Arg` instance to test certain hoisting functions I haven't uploaded yet, which I don't like, and also force us to write `Eq1` and (worse) `Show1` instances for `TestF`.

I still have a little ways to go before I can test everything I want, but I think I'm getting close.

dagit commented:
That's a clever idea. I was definitely going to go more primitive than that and just write tests for things like `SeqT IO Int` and `Seq Int`, but I do like the idea of having the tests work for arbitrary types.

And if you prefer `QuickCheck` it's totally fine to use that. With `hspec` you can easily combine different testing approaches/frameworks.

dagit commented:
I pushed some minor tweaks to the cabal file and the CI config to workaround some silly build issues. So I would recommend rebasing on `main` when you work on this next.

treeowl commented:
Okay....

On Mon, Jul 26, 2021, 5:38 PM Jason Dagit ***@***.***> wrote:

> I pushed some minor tweaks to the cabal file and the CI config to
> workaround some silly build issues. So I would recommend rebasing on main
> when you work on this next.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-887045178>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7J6M2FE2P2F6U36QHLTZXIT7ANCNFSM5A53S64Q>
> .
>

treeowl commented:
BTW, the challenge I'm dealing with now is getting sensible size ranges for testing. Using the general tools I've found in hedgehog (e.g. `recursive`), I get a bunch of trivial test cases and/or occasional *enormous* ones that blow up the test time. So I think I want to choose the exact (or nearly exact) number of nodes *first* and thread everything through. It feels very, very low level. Note: this isn't any easier with QuickCheck.

dagit commented:
I think in hedgehog land people write multiple generators based on
strategies for generating a value. The way I’ve used it in the past when
generating ASTs was to make generators that created ASTs with certain
properties like avoiding a particular node type and that sort of thing.

And yes, my experience with quick check and tree generation is that you
usually have to take the size as a non randomized parameter and then do
like a list comprehension to test a set of hand picked sizes. And this is
just because trees tend to grow exponentially in the simple cases of
generating them. If the size was the number of nodes it might be different
but it’s also harder to think about generating a tree like that.

On Mon, Jul 26, 2021 at 15:08 David Feuer ***@***.***> wrote:

> BTW, the challenge I'm dealing with now is getting sensible size ranges
> for testing. Using the general tools I've found in hedgehog (e.g.
> recursive), I get a bunch of trivial test cases and/or occasional
> *enormous* ones that blow up the test time. So I think I want to choose
> the exact (or nearly exact) number of nodes *first* and thread everything
> through. It feels very, very low level. Note: this isn't any easier with
> QuickCheck.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-887060025>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC547INHVWBTKXCJNNKQLLTZXMEJANCNFSM5A53S64Q>
> .
>

treeowl commented:
I think the first thing I want is something like

```haskell
splat :: Int -> Gen [Int]
```

that takes a non-negative integer and produces an arbitrary-length list of positive integers that sum to it. Then I can write functions like

```haskell
blahSized :: MonadGen m => (Int -> m a) -> Int -> m (Blah a)
```

Given a way to produce `a`s of *roughly* any given size, produce a `Blah a` of roughly the given size.

I seem to remember doing something somewhat similar to thisb years back, but I don't really remember where. `containers`, possibly?

treeowl commented:
@dagit, I've added hoists here, and a good start to a test suite. I haven't added `MonadReader` or `MonadError` tests yet. I can certainly add a couple of `MonadReader` unit tests tonight if you want to wait, but I don't yet know how to test it more generally, and I really have no idea what to look for from the `MonadError` instance here. I could probably throw in a few unit tests too, but I don't know which are the "hard" cases.

dagit commented:
I’d say add anything that is quick to do and document the rest with issues
so that we don’t lose track of them. Maybe even update your todo comments
to cite the issues(a) you create.

Thanks!

On Tue, Jul 27, 2021 at 12:11 David Feuer ***@***.***> wrote:

> @dagit <https://github.com/dagit>, I've added hoists here, and a good
> start to a test suite. I haven't added MonadReader or MonadError tests
> yet. I can certainly add a couple of MonadReader unit tests tonight if
> you want to wait, but I don't yet know how to test it more generally, and I
> really have no idea what to look for from the MonadError instance here. I
> could probably throw in a few unit tests too, but I don't know which are
> the "hard" cases.
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-887765171>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC547PHPOIN6UPANII4HOLTZ4AFHANCNFSM5A53S64Q>
> .
>

treeowl commented:
I added unit tests for the `MonadReader` instance. I don't really know how to test the `MonadError` instance, but I'll open an issue to do so.

dagit commented:
Do you want me to review it and possible merge it or do you want to keep working on it?

treeowl commented:
Please review for merge.

On Wed, Jul 28, 2021, 11:16 PM Jason Dagit ***@***.***> wrote:

> Do you want me to review it and possible merge it or do you want to keep
> working on it?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-888768440>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7JHD47FLU6AYP6OFV3T2DBY5ANCNFSM5A53S64Q>
> .
>

dagit commented on logict-sequence.cabal:
```diff
@@ -51,7 +53,8 @@ library
     build-depends: type-aligned >= 0.9.6 && < 0.10
     build-depends: sequence >= 0.9.8 && < 0.10
-    build-depends: logict
+    build-depends: logict >= 0.7.1.0
```
I wonder if we should also have an upper bound like maybe `< 0.8`, or didn't cabal get some fancy new operator for bounds like `==^*&&@@<>` (I really can't remember what it was, sorry).

treeowl replied:
You're asking me? That's semi-hilarious. That does seem to be popular, though.

dagit commented on src/Control/Monad/Logic/Sequence/Internal.hs:
```diff
@@ -305,6 +325,87 @@ observeAll :: Seq a -> [a]
+fromLogicT :: Monad m => L.LogicT m a -> SeqT m a
+fromLogicT (L.LogicT f) = fromView $ f (\a v -> return (a :< fromView v)) (return Empty)
+
+instance MFunctor SeqT where
```
I'm totally unfamiliar with `MFunctor` so I'm just going to take your word on this stuff.

treeowl replied:
`MFunctor` is pretty simple in practice. It's used to lift monad morphisms (e.g., `lift`, or of course various *isomorphisms* between monads) the way `fmap` is used to lift functions. Have a `SeqT m a` and need a `SeqT (MaybeT m) a`? `hoist lift` will convert it. Etc.

dagit commented on test/Test.hs:
```diff
@@ -1,7 +1,191 @@
+-- chosen in advance to make sure the target is met.
+
+-- | Generate a partition of a non-negative integer into positive
+-- integers. This is not statistically fair because I'm not that smart.
```
I would argue that if simplistic random values aren't cutting it, then you probably want to use more foundational tools for your correctness like proving properties about a model of the monad in Coq. So I wouldn't sweat it about things like fairness.

treeowl replied:
Eh, fairness helps with test coverage, but I'll try not to worry *too* much.

dagit commented on test/Test.hs:
```diff
@@ -1,7 +1,191 @@
+      s <- forAll simpleSeqT
+      hoist (\x -> x) s === s
+
+{-
```
Is there any reason to keep this commented code? My preference would be to deleted it.

treeowl replied:
Yeah, I can do that.

treeowl commented on src/Control/Monad/Logic/Sequence/Morph.hs:
```diff
@@ -0,0 +1,9 @@
+  ( hoistUnexposed
+  , hoistPost
+  , hoistPostUnexposed
+  , hoist
```
I'm a little uncomfortable about re-exporting the `hoist` method. Should I

1. Export a type-specialized one instead?
2. Rename `hoistUnexposed` to `hoistPreUnexposed` and export a type-specialized `hoist` called `hoistPre`?

I'm leaning towards 2. Note: the `Unexposed` terminology is stolen from `streaming`.

dagit replied:
I'll be honest. I don't know much about this hoist stuff or how often anyone will need it. Are you uncomfortable because it's it can be unsafe in some context or just trying not to pollute things when people import stuff?

dagit reviewed (changes requested):
Thanks! Overall this is a great start to a testsuite.

Two minor things, the upperbound on `logict` and the commented code at the end of Test.hs.

treeowl commented on test/Test.hs:
```diff
@@ -1,7 +1,191 @@
+-- some examples being unmanageably enormous. So I've decided to
+-- go with a "nuclear option". First, I select the exact total number
+-- of nodes in the SeqT. Then at each stage, the exact total size is
+-- chosen in advance to make sure the target is met.
```
I shouldn't say "exact" because I'm not so sure it is. "Approximate" is plenty good enough.

treeowl commented:
Pollution, pollution.... And all that jazz. Don't want someone to add an
import of our module and suddenly have a redundant import warning for
`Control.Monad.Morph`.

On Wed, Jul 28, 2021 at 11:53 PM Jason Dagit ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In src/Control/Monad/Logic/Sequence/Morph.hs
> <https://github.com/dagit/logict-sequence/pull/25#discussion_r678802780>:
>
> > @@ -0,0 +1,9 @@
> +module Control.Monad.Logic.Sequence.Morph
> +  ( hoistUnexposed
> +  , hoistPost
> +  , hoistPostUnexposed
> +  , hoist
>
> I'll be honest. I don't know much about this hoist stuff or how often
> anyone will need it. Are you uncomfortable because it's it can be unsafe in
> some context or just trying not to pollute things when people import stuff?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#discussion_r678802780>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAOOF7PYLGY6DQ7MJEDDMILT2DGDXANCNFSM5A53S64Q>
> .
>

treeowl commented on test/Test.hs:
```diff
@@ -1,7 +1,191 @@
+simpleSeqT :: MonadGen m => m (SeqT TestM Int)
+simpleSeqT = genSeqT (Gen.integral $ Range.constant 0 5)
+
+simpleSeqT' :: MonadGen m => m (Size, SeqT TestM Int)
```
I don't think this is used anywhere. I just put it together to debug my test case size issues.

treeowl commented:
I think I've addressed your review, and my own concerns. I added some more property tests too, including all the `Alternative` and `MonadPlus` laws (with left distribution).

treeowl commented:
Uh oh.... The `Read` instance doesn't work. I'm investigating now.

treeowl commented:
Er... Sorry, it might. The problem might be the test I'm working on for it. Stand by...

treeowl commented:
Yeah, it's the test. Don't worry about it for now.

treeowl commented:
OK... I simplified the instance situation in the test suite so I could derive more stuff, and now `read . show` works as it should. I also added a few lifted instances for `SeqT`, which should be tested eventually.

treeowl commented:
OK, looks like I got all the compat crud sorted. Could you possibly review again? I want to see this merged!

dagit commented:
Awesome. I’ll try to find time tonight, but by this weekend at the latest.

On Thu, Jul 29, 2021 at 15:13 David Feuer ***@***.***> wrote:

> OK, looks like I got all the compat crud sorted. Could you possibly review
> again? I want to see this merged!
>
> —
> You are receiving this because you were mentioned.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/dagit/logict-sequence/pull/25#issuecomment-889495708>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC547K625G4UVI25AJALU3T2HHCHANCNFSM5A53S64Q>
> .
>

treeowl commented:
Hrmm... I tried to get the `Data.Functor.Classes` stuff working for `base < 4.9.0` too, but for reasons I don't really understand that's been a no-go. If you want, you can try to figure out how to juggle the versions of `base` and `transformers`, with the change in location of the module and the change in the class definition, but I'm not really in the mood to deal with it.

dagit commented:
Thanks once again!

treeowl commented:
Yaaaaay!!!
